contracts/core/Allo.sol

```diff
@@ -3,28 +3,52 @@ pragma solidity 0.8.19;
 
 import { Metadata } from "./libraries/Metadata.sol";
 import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import "../interfaces/IAllocationStrategy.sol";
+import "../interfaces/IDistributionStrategy.sol";
+import "./Registry.sol";
+
 
 contract Allo is Initializable {
-    struct PoolData {
-        address identity;
-        address allocationStrategy;
-        address distributionStrategy;
+
+     error NO_ACCESS_TO_ROLE();
+
+    /// @notice Struct to hold details of an Pool
+    struct Pool {
+        bytes32 identityId;
+        IAllocationStrategy allocationStrategy;
+        IDistributionStrategy distributionStrategy;
         Metadata metadata;
         bool active;
     }
 
+    /// @notice Registry of pool creators
+    Registry public registry;
+
+    /// @notice Incremental Index
+    uint private _poolIndex;
+
+    /// @notice Pool.id -> Pool
+    mapping(uint256 => Pool) public pools;
+
     /**
      * @notice Initializes the contract after an upgrade
      * @dev In future deploys of the implementation, an higher version should be passed to reinitializer
      */
-    function initialize() public reinitializer(1) {}
+    function initialize(address _registry) public reinitializer(1) {
+        registry = Registry(_registry);
+        // ASK: should there be an update function to update registry ? Ownable contract ?
+        // Should this contract be upgradable
```

Comment:
@codenamejason @zobront @KurtMerbeth

Reply:
I'm totally in favor of having the contracts upgradeable. At least at the beginning.

Will we have only one registry, or a list of "trusted sources? (E.g. registry and registry relayer)

Reply:
Ah, good question. 
I want to say 1 registry cause now you create your identityId and same anchor on every chain. Hence the need to support multiple Allo registries ceases to exist (or am I wrong ?)

Reply:
Does this mean one registry per chain? or do we have a base registry on a single chain and the front-ends just call the registry on mainnet (PGN) no matter what chain they are on?

Reply:
I think we'll discuss that on the call today but TLDR:
Since it's possible to port your identity (anchor + identityIndex) by controlling the nonce + msg.sender + name
We don't have to worry about whitelisting multiple registries 


>  Does this mean one registry per chain
Yes. If someone wanted to deploy their own duplicate registry, they could do that and our system would still work cause identity can be ported

Reply:
Decided that this is the only contract which we'll make upgradeable. Later on once we are stable -> we'll set the update perms to address(0) and freeze upgradability.